pkg/raft/log.go

```diff
@@ -529,6 +506,95 @@ func (l *raftLog) scan(lo, hi uint64, pageSize entryEncodingSize, v func([]pb.En
 // The returned slice can be appended to, but the entries in it must not be
 // mutated because they can be still shared with the raftLog/unstable.
 func (l *raftLog) slice(lo, hi uint64, maxSize entryEncodingSize) ([]pb.Entry, error) {
+	snap := l.snap()
+	return snap.slice(lo, hi, maxSize)
+}
+
+// l.firstIndex()-1 <= lo <= hi <= l.lastIndex().
+func (l *raftLog) mustCheckOutOfBounds(lo, hi uint64) error {
+	snap := l.snap()
+	return snap.mustCheckOutOfBounds(lo, hi)
+}
+
+func (l *raftLog) zeroTermOnOutOfBounds(t uint64, err error) uint64 {
+	if err == nil {
+		return t
+	}
+	if err == ErrCompacted || err == ErrUnavailable {
+		return 0
+	}
+	l.logger.Panicf("unexpected error (%v)", err)
+	return 0
+}
+
+func (l *raftLog) snap() LogSnapshot {
+	return LogSnapshot{
+		first:    l.firstIndex(),
+		unstable: l.unstable.logSlice,
+		storage:  l.storage,
+		logger:   l.logger,
+	}
+}
+
+// LogSnapshot encapsulates the point-in-time state of the raft log.
+type LogSnapshot struct {
+	first    uint64
+	unstable logSlice
+	storage  Storage
```

Comment:
It's unexpected to me that the `Storage` is part of the snapshot. The snapshot provides a "view" into the Storage, assuming the `Storage` isn't mutated. But the snapshot doesn't "snapshot" the `Storage`. So it's worth considering whether all the methods on a `LogSnapshot` that actually need to do IO should take the `Storage` as a parameter.

Reply:
(That said, if this is all complicated to disentagle, what we have here is fine and since we need to pass this `LogSnapshot` around, it could be difficult to do the plumbing for `Storage`).

Reply:
I wonder if we could make a `Storage` wrapper which asserts that is wasn't mutated. E.g. I could imagine incrementing some (in memory only) "epoch" int in it whenever we mutate the log storage. Then make a lean wrapper around our `Storage` impl which, on every call, asserts that the epoch hasn't changed.

```golang

type Storage interface {
	// everything we have today in Storage
}

type StorageRO = Storage

type StorageRW interface {
	Storage
	Snapshot() StorageRO
}


// somewhere in replica_raftstorage.go
func (r *replicaRaftStorage) Snapshot() raft.StorageRO {
	return wrapper{r: r, epoch: r.epoch}
}
```

Reply:
We are faking storage snapshots but just requiring that storage is not mutated for a while.

Reply:
There might be some confusion between the "log snapshot" and the "state machine snapshot". Probably a good time to complete [this](https://github.com/cockroachdb/cockroach/blob/83a68a474a31b6ffc2d7c45d0097b859568e60dc/pkg/raft/storage.go#L50) TODO.